### PR TITLE
Refactors get_bucket_policy and set_bucket_policy apis

### LINKED ISSUE #635: set/getBucketPolicy should take/return policy JSON file
Currently the set/getBucketPolicy APIs are high level providing fine grained access. It was decided to make set/getBucketPolicy a low level API directly dealing with the policy json file to make the client libs simpler.

Changes similar to https://github.com/minio/minio-js/pull/676 is needed.

### PR DESCRIPTION
**Pull request by ebozduman**
Fixes #635

### CODE REVIEW

File: examples/set_bucket_policy.py [hunk 1]
```diff
@@ -19,32 +19,114 @@
...
+    }
+    client.set_bucket_policy('my-bucketname', json.dumps(policy_READ_ONLY))
 
-    # Set policy Policy.READ_WRITE to bucket 'my-bucketname' and
-    # prefix 'public-folder/' which enables
-    # 'my-bucketname/public-folder/' read/writeable by everyone.
-    client.set_bucket_policy('my-bucketname', 'public-folder/',
-                             Policy.READ_WRITE)
+    # Set bucket policy to read-write for bucket 'my-bucketname'
+    policy_READ_WRITE = {
```
Comment (poornas) [inline examples/set_bucket_policy.py]: s/policy_READ_WRITE/policy_read_write

File: examples/set_bucket_policy.py [hunk 2]
```diff
@@ -19,32 +19,114 @@
...
+        ]
+    }
+    client.set_bucket_policy('my-bucketname', json.dumps(policy_READ_WRITE))
 
-    # Set policy Policy.WRITE_ONLY to bucket 'my-bucketname' and
-    # prefix 'incoming' which enables 'my-bucketname/incoming'
-    # writeable by everyone.
-    client.set_bucket_policy('my-bucketname', 'incoming',
-                             Policy.WRITE_ONLY)
+    # Set bucket policy to write-only for bucket 'my-bucketname'
```
Comment (poornas) [inline examples/set_bucket_policy.py]: s/policy_WRITE_ONLY/policy_write_only

File: tests/functional/tests.py [hunk 3]
```diff
@@ -1414,46 +1413,97 @@ def threaded_get_object(cli, bckt_name, obj_name, obj_data):
...
+                }
+            ]
+        }
+        # Set read-only policy
+        client.set_bucket_policy(bucket_name, json.dumps(policy))
         # Validate if the policy is set correctly
-        policy_name = client.get_bucket_policy(bucket_name)
-        if policy_name != Policy.READ_ONLY:
-            raise ValueError('Failed to set ReadOnly bucket policy: ' + policy_name)
+        if not policy_validated(client, bucket_name, policy):
```
Comment (poornas) [inline tests/functional/tests.py]: wouldn't a simple string compare of policy json to be uploaded and policy retrieved by get_bucket_policy validate this? Don't see a need for this helper function.

File: docs/API.md [hunk 4]
```diff
@@ -311,56 +311,75 @@ for obj in uploads:
...
 
 __Example__
 
 
 ```py
-# Get current policy of all object paths in bucket that begin with my-prefixname.
-policy = minioClient.get_bucket_policy('mybucket',
-                                       'my-prefixname')
+# Get current policy of all object paths in bucket "mybucket".
+policy = minioClient.get_bucket_policy('mybucket')
```
Comment (vadmeste) [inline docs/API.md]: When the bucket doesn't have a policy, this function raises minio.error.NoSuchBucketPolicy exception, before, we were returning None. It makes sense to discuss if we need to keep the old behavior or not, and document this in all cases.

Comment (harshavardhana) [reply docs/API.md]: We should return `NoSuchBucketPolicy` since that would be expected now. it's a breaking change so don't expect old behavior.

Comment (vadmeste) [reply docs/API.md]: okay, then we need to document this

File: tests/functional/tests.py (hunk 3 above)
Comment (ebozduman) [reply tests/functional/tests.py]: I've seen different, sometimes duplicate `Action` and/or `Resource` entries returned from the server. So, the format of the json response is not guaranteed. That's why I've introduced the helper function.

File: examples/set_bucket_policy.py (hunk 2 above)
Comment (ebozduman) [reply examples/set_bucket_policy.py]: Done.

File: examples/set_bucket_policy.py (hunk 1 above)
Comment (ebozduman) [reply examples/set_bucket_policy.py]: Done.

File: examples/set_bucket_policy.py [hunk 5]
```diff
@@ -19,32 +19,114 @@
...
                access_key='YOUR-ACCESSKEYID',
                secret_key='YOUR-SECRETACCESSKEY')
 
 # Make a new bucket
 try:
-    # Set policy Policy.READ_ONLY to bucket 'my-bucketname' which
-    # enables 'my-bucketname' readable by everyone.
-    client.set_bucket_policy('my-bucketname', '', Policy.READ_ONLY)
+    # Set bucket policy to read-only for bucket 'my-bucketname'
+    policy_READ_ONLY = {
```
[the remark this answers is not in the document]
Comment (ebozduman) [reply examples/set_bucket_policy.py]: Done.

File: docs/API.md [hunk 6]
```diff
@@ -311,56 +311,75 @@ for obj in uploads:
...
+                        "Sid":"",
+                        "Effect":"Allow",
+                        "Principal":{"AWS":"*"},
+                        "Action":"s3:GetObject",
+                        "Resource":"arn:aws:s3:::mybucket/*"
+                        }
+                    ]}
+
+
+minioClient.set_bucket_policy('mybucket', policy_READ_ONLY)
```
[the remark this answers is not in the document]
Comment (ebozduman) [reply docs/API.md]: Done.

File: docs/API.md (hunk 4 above)
Comment (ebozduman) [reply docs/API.md]: You mean in the Release Notes, right? Yes.

**Comment by ebozduman** [comment]
@vadmeste, @harshavardhana, @poornas ,
Could you re-review it? Thanks.

**Comment by poornas** [comment]
@ebozduman, please check the build failures.

File: docs/API.md [hunk 7]
```diff
@@ -317,55 +317,75 @@ for obj in uploads:
...
 
 __Example__
 
 
 ```py
-# Set policy Policy.READ_ONLY to all object paths in bucket that begin with my-prefixname.
-minioClient.set_bucket_policy('mybucket',
-                              'my-prefixname',
-                              Policy.READ_ONLY)
+# Set policy to read only to all object paths in bucket.
```
[the remark this answers is not in the document]
Comment (ebozduman) [reply docs/API.md]: Done.
Also resolved Travis problem.

**Comment by ebozduman** [comment]
@poornas, @harshavardhana, @krishnasrinivas,
Could you review it again?
Thanks.

**Comment by ebozduman** [comment]
@poornas, @harshavardhana, @krishnasrinivas,
Ping! for review...

**Pull request merged**